<a href="https://colab.research.google.com/github/tlvhoang06/midterm-lab-combinatorics/blob/main/MidTerm_Lab_ipynb_e593e0d0e25e47567b593818d4936445_txt.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# BỘ ĐỀ ÔN LUYỆN THỰC HÀNH MÔN TOÁN HỌC TỔ HỢP

## Cấu trúc chung

Đây là bộ đề ôn luyện thi thực hành hoàn chỉnh gồm 20 câu. Sử dụng các kỹ thuật lập trình Python đã học (sử dụng vòng lặp có điều kiện, các hàm $math$, $itertools$, và $sympy.FiniteSet$).

Mỗi câu trả lời được trình bày theo cấu trúc 3 phần yêu cầu:
- Code đếm bằng vòng lặp/hàm Python cơ bản để liệt kê và tính toán ra lời giải,
- Lời giải bằng Toán có giải thích,
- Code so sánh kết quả.


Gợi ý: Phần code sử dụng itertools.permutations, itertools.combinations, ... hoặc for loop để tạo liệt kê. Sử dụng math.factorial, math.comb ... để tính giá trị công thức như đã học trên lớp thực hành

## Chủ đề 1: Các quy tắc đếm cơ bản

### Câu ví dụ: Đếm bội số

### Có bao nhiêu bội số của 7 trong các số nguyên từ 1 đến 1001?

#### Lời giải ví dụ:

##### 1. Code đếm lời giải (Vòng lặp có điều kiện)

In [8]:
N = 1001
divisor = 7
count_loop = 0
for i in range(1, N + 1):
    if i % divisor == 0:
        count_loop += 1
print(f"Code đếm bằng vòng lặp: {count_loop}")


Code đếm bằng vòng lặp: 143


##### 2. Lời giải bằng Toán, có giải thích

Đây là bài toán đếm số phần tử trong một dãy số chia hết cho k

- Công thức: Số bội số của k trong khoảng $[1,N]$ là $\lfloor \frac{n}{k} \rfloor$.
- Giải thích: Ta cần tìm số nguyên sao $m$ cho $7m <= 1001$

$$ m = \lfloor \frac{1001}{7} \rfloor = 143$$

##### 3. Code so sánh lời giải

In [9]:
import math
N = 1001
divisor = 7
count_math = math.floor(N / divisor)
print(f"Code đếm bằng vòng lặp: {count_loop}")
print(f"Code sử dụng công thức: {count_math}")
print(f"Kết quả so sánh: {count_loop == count_math}")


Code đếm bằng vòng lặp: 143
Code sử dụng công thức: 143
Kết quả so sánh: True


### Câu 1:

### Có bao nhiêu số chẵn từ 100 đến 999 có các chữ số đôi một khác nhau?

## **+ Gọi "ABC" là số từ 100 đến 999**
### - Cách chọn A : 1 -> 9
### - Cách chọn B : 1 -> 9
### - Cách chọn C : 0, 2, 4, 6, 8

## **+ TH1: C = 0**
 ### C : 1 cách chọn
 ### B : 9 cách chọn
 ### A : 8 cách chọn
 ### --> có 1 x 9 x 8 = 72 cách
## **+ TH2: C != 0**
### C : 4 cách chọn
### A : 8 cách chọn
### B : 8 cách chọn
### --> có 4 x 8 x 8 = 256 cách
## **==> Vậy có tổng cộng 256 + 72 = 328 cách chọn**

In [19]:
count = 0
for i in range(100, 999):
  if i % 2 == 0:
    s = str(i)
    if(s[0] != s[1] and s[1] != s[2] and s[0] != s[2]):
      count = count + 1

So cac so chan doi mot khac nhau tu 100 den 999 la: 328


### Câu 2:

### Bao nhiêu số lẻ 5 chữ số có chính xác một cặp chữ số trùng (tất cả chữ số khác đều khác nhau), và chữ số cuối là lẻ?

# **Giải bằng toán học**
## + Gọi số cần tìm là abcde
## + Ràng buộc:
## - e có 5 cách chọn: 1, 3, 5, 7, 9
## - a có 9 cách chọn: 1 -> 9
## + Xét: Vị trí trùng nằm ở e
## TH1: a = e
## a và e có 5 cách chọn (thỏa a != 0)
## 3 vị trí còn lại có 9P3 cách chọn
## -> Có 5 * 9P3 = 2520 (Cách)
## TH2:



In [17]:
count = 0
for i in range(10000, 100000):
    if i % 2 == 0:
        continue
    s = str(i)
    valid = True
    pair = 0
    for j in range(len(s)):
        freq = s.count(s[j])
        if freq > 2:
            valid = False
            break
        if freq == 2:
             pair += 1
    if valid and pair == 2:
        count += 1
print(count)


22680


## Chủ đề 2: Hoán vị, Chỉnh hợp, Tổ hợp


### Câu 3:
### Từ các ký tự $ {a,b,c,d,e,f} $, có bao nhiêu cách tạo ra một chuỗi gồm 3 ký tự tùy ý?

### Câu 4:

### Từ các ký tự $ {a,b,c,d,e,f} $, có bao nhiêu cách tạo ra một chuỗi gồm 3 ký tự đôi một khác nhau??

### Câu 5:

### Một câu lạc bộ cờ vua có 12 thành viên. Có bao nhiêu cách chọn ra một đội gồm 5 người để tham gia giải đấu (không quan tâm thứ tự)?

### Câu 6:

### Có bao nhiêu chuỗi nhị phân chiều dài 8 chứa đúng 3 bit 0?

### Câu 7:

### Trong một giải chạy marathon có 20 vận động viên tham gia. Có bao nhiêu khả năng khác nhau cho 3 vị trí Nhất, Nhì, và Ba?

### Câu 8:

### Từ một nhóm gồm 8 nhà khoa học nam và 6 nhà khoa học nữ, có bao nhiêu cách thành lập một ủy ban gồm 6 người, sao cho ủy ban có ít nhất 3 nhà khoa học nữ?

### Câu 9:

### Có bao nhiêu cách sắp xếp 8 ký tự của từ VISITING? (I lặp 3 lần, V, S, T, N, G lặp 1 lần)

### Câu 10:

### Có bao nhiêu cách sắp xếp 6 quyển sách Toán khác nhau và 4 quyển sách Hóa học khác nhau lên một kệ sách dài nếu các sách cùng một môn phải được xếp cạnh nhau?

### Câu 11:

### Một người đi mua 12 quả trái cây từ một cửa hàng có bán 5 loại: táo, cam, chuối, lê, và xoài. Hỏi có bao nhiêu cách chọn khác nhau? (Giả sử cửa hàng có đủ số lượng mỗi loại).

### Câu 12:

### Có bao nhiêu cách chia 10 lá thư giống hệt nhau vào 4 hộp thư khác nhau?

### Câu 13:

### 5 cặp vợ chồng đi dự đám cưới được xếp ngồi vào một bàn tròn sao cho các ông chồng ngồi cạnh nhau và các bà vợ ngồi cạnh nhau.

### Câu 14:

### Có bao nhiêu cách chọn ít nhất một trái từ 6 trái táo (giống nhau) và 8 trái cam (giống nhau)?

### Câu 15:

### Một tiệm kem có 10 loại kem khác nhau. Một khách hàng muốn mua 3 loại kem khác nhau trong đó phải có kem dừa.

## Chủ đề 3: Nguyên lý Bù trừ

### Câu 16:

### Có bao nhiêu số nguyên từ 1 đến 100 chia hết cho 2, 3 hoặc 5?

### Câu 17:

### Một hiệu bánh có 30 loại bánh khác nhau (18 mặn, 12 ngọt). Một khách hàng muốn chọn 5 loại bánh khác nhau sao cho đội hình gồm cả 2 loại mặn và ngọt?

### Câu 18:

### Mã PIN gồm 4 chữ số (0-9). Mã PIN hợp lệ phải có ít nhất một chữ số lớn hơn 5 (≥6) và ít nhất một chữ số nhỏ hơn hoặc bằng 5 (≤5). Hỏi có bao nhiêu mã PIN hợp lệ?

### Câu 19:

### Một nhân viên kho có 6 nhãn vận chuyển cho 6 thùng hàng riêng biệt. Anh ta dán nhãn một cách ngẫu nhiên. Có bao nhiêu cách dán nhãn để ít nhất một thùng hàng được dán đúng nhãn của nó?

### Câu 20:

### Trong các số (1..200), có bao nhiêu số chia hết cho 3 hoặc 7 hoặc 9?